hummingbot/connector/exchange/binance/binance_api_user_stream_data_source.py

```diff
@@ -62,57 +61,58 @@ async def get_listen_key(self):
-                    response: aiohttp.ClientResponse = response
-                    if response.status != 200:
-                        raise IOError(f"Error fetching user stream listen key. Response: {response}")
-                    data: Dict[str, str] = await response.json()
-                    return data["listenKey"]
-
-    async def ping_listen_key(self, listen_key: str) -> bool:
+                response = await client.post(url=url,
+                                             headers={"X-MBX-APIKEY": self._binance_client.API_KEY})
+
+                if response.status != 200:
+                    raise IOError(f"Error fetching user stream listen key. Response: {response}")
+                data: Dict[str, str] = await response.json()
+                return data["listenKey"]
+
+    async def ping_listen_key(self) -> bool:
         async with aiohttp.ClientSession() as client:
             async with self._throttler.execute_task(limit_id=CONSTANTS.BINANCE_USER_STREAM_PATH_URL):
                 url = binance_utils.private_rest_url(path_url=CONSTANTS.BINANCE_USER_STREAM_PATH_URL, domain=self._domain)
-                async with client.put(url=url,
-                                      headers={"X-MBX-APIKEY": self._binance_client.API_KEY},
-                                      params={"listenKey": listen_key}) as response:
-                    data: Tuple[str, any] = await response.json()
-                    if "code" in data:
-                        self.logger().warning(f"Failed to refresh the listen key {listen_key}: {data}")
-                        return False
-                    return True
-
-    async def _iter_messages(self, ws: websockets.WebSocketClientProtocol) -> AsyncIterable[str]:
-        # Terminate the recv() loop as soon as the next message timed out, so the outer loop can reconnect.
+                response = await client.put(url=url,
+                                            headers={"X-MBX-APIKEY": self._binance_client.API_KEY},
+                                            params={"listenKey": self._current_listen_key})
+                data: Tuple[str, any] = await response.json()
+                if "code" in data:
+                    self.logger().warning(f"Failed to refresh the listen key {self._current_listen_key}: {data}")
+                    return False
+                return True
+
+    async def _create_websocket_connection(self) -> aiohttp.ClientWebSocketResponse:
+        """
+        Initialize WebSocket client for UserStreamDataSource
+        """
+        try:
+            url = f"{CONSTANTS.WSS_URL.format(self._domain)}/{self._current_listen_key}"
+            self.logger().info(f"Connecting to {url}.")
+            return await aiohttp.ClientSession().ws_connect(url=url,
+                                                            heartbeat=30.0,
+                                                            autoping=False)
```

Comment:
Is it correct that the autoping for the user stream websocket is disabled but enabled for the order book stream websocket? What is the heartbeat used for if autoping is disabled?

Reply:
Heartbeat is used to send ping frames to the server to determine if the connection is still up. While, `autoping` responds with a pong frame upon receiving a ping frame.